packages/metro-resolver/src/errors/formatFileCandidates.js

```diff
@@ -11,15 +11,53 @@
 
 'use strict';
 
+import path from 'path';
+
 import type {FileCandidates} from '../types';
 
-function formatFileCandidates(candidates: FileCandidates): string {
+function formatFileCandidates(
+  candidates: FileCandidates,
+  allowIndex: boolean = false,
+): string {
   if (candidates.type === 'asset') {
     return candidates.name;
   }
   let formatted = candidates.filePathPrefix;
+
+  if (allowIndex) {
+    formatted += `(${path.sep}index)?`;
+  }
+
   if (candidates.candidateExts.length) {
-    formatted += '(' + candidates.candidateExts.filter(Boolean).join('|') + ')';
+    const exts = candidates.candidateExts.filter(Boolean);
```

Comment:
How does this logic handle entries in `sourceExts` that contain a `.`? Not sure if there's an existing test for this, but internally at Meta we have a config with `.fb.js` as a possible extension. Would this logic split it into `fb` and `js`, suggesting that other compound extensions like `.fb.ts` are included (or given special precedence) when they're not? Or would it correctly preserve `.fb.js` as a unit? (Caveat: I'm too jetlagged to try to follow the code right now.)

Reply:
This potentially introduces a three-level deep extension (`.native.fb.js`) which could be formatted like:

```
* missing(/index)?(.native)?(.fb)?.(js|json)
```

- `missing.native.fb.js`
- `missing.fb.js`
- `missing.native.js`
- `missing.js`
- `missing/index.native.fb.js`
- `missing/index.fb.js`
- `missing/index.native.js`
- `missing/index.js`


Does this look right to you?